### PR TITLE
Add Display implementation to DebugName.

### PR DESCRIPTION
**Pull request by gagnus**
# Objective

- When writing "in game" debugging tools, quite often you need the name of an entity (for example an entity tree). DebugName is the usual way of doing that.
- A recent change to Entity's Debug implementation meant it was no longer a minimal {index}v{generation} but instead a more verbose auto generated Debug.
- This made DebugName's Debug implementation also verbose

## Solution

- I changed DebugName to derive Debug automatically and added a new (preferred) Display implementation for it which is the preferred name for an entity. If the entity has a Name component its the contents of that, otherwise it is {index}v{generation} (though this does not use Display of the Entity as that is more verbose than this).

## Testing

- I've added a new test in name.rs which tests the Display implementation for DebugName by using to_string.

---

## Migration Guide

- In code which uses DebugName you should now use the Display implementation rather than the Debug implementation (ie {} instead of {:?} if you were printing it out).

### CODE REVIEW

**Comment by alice-i-cecile** [comment]
> Additionally (more controversial maybe) I have returned the Display implementation in Entity to just be {index}v{generation} rather than also having |{bits} on (where {bits} is a 64-bit decimal version of generation << 32 | index).

I don't like this change personally, but we should definitely split this off to make sure the other work doesn't get blocked.

**Comment by gagnus** [comment]
I've updated to no longer change the Display implementation for Entity and instead use {index}v{generation} directly in DebugName.

File: crates/bevy_core/src/name.rs [hunk 1]
```diff
@@ -100,26 +100,31 @@ impl std::fmt::Debug for Name {
...
+///             // for entities which don't have a Name.
+///             // You can still use the Debug impl it is just quite verbose.
+///             bevy_utils::tracing::error!("Score for {} is invalid", name);
 ///         }
 ///     }
 /// }
 /// # bevy_ecs::system::assert_is_system(increment_score);
 /// ```
 #[derive(QueryData)]
+#[query_data(derive(Debug))]
```
Comment (alice-i-cecile) [inline crates/bevy_core/src/name.rs:114]: Wow, I had no idea this was possible. Neat! It's even well documented.

**Review by alice-i-cecile — APPROVED** [review]

File: crates/bevy_core/src/name.rs [hunk 2]
```diff
@@ -100,26 +100,31 @@ impl std::fmt::Debug for Name {
 ///     for (name, mut score) in &mut scores {
 ///         score.0 += 1.0;
 ///         if score.0.is_nan() {
-///             bevy_utils::tracing::error!("Score for {:?} is invalid", name);
+///             // use the Display impl to return either the Name
+///             // where there is one, or {index}v{generation}
+///             // for entities which don't have a Name.
+///             // You can still use the Debug impl it is just quite verbose.
+///             bevy_utils::tracing::error!("Score for {} is invalid", name);
```
Comment (janhohenheim) [inline crates/bevy_core/src/name.rs:107]:
```suggestion
///             bevy_utils::tracing::error!("Score for {name} is invalid");
```

File: crates/bevy_core/src/name.rs [hunk 3]
```diff
@@ -100,26 +100,31 @@ impl std::fmt::Debug for Name {
 ///     for (name, mut score) in &mut scores {
 ///         score.0 += 1.0;
 ///         if score.0.is_nan() {
-///             bevy_utils::tracing::error!("Score for {:?} is invalid", name);
+///             // use the Display impl to return either the Name
```
Comment (janhohenheim) [inline crates/bevy_core/src/name.rs:103]: This kind of implementation information seems a bit too in-depth for an example. Additionally, the part about debug being more verbose is imo implicitly to be expected.

Suggestion: move this information outside of the example (maybe a section `# Implementation`?) and remove references to `Debug`

**Review by janhohenheim — CHANGES_REQUESTED** [review]
Good change, just some minor nitpicks :)

File: crates/bevy_core/src/name.rs [hunk 4]
```diff
@@ -198,3 +203,24 @@ impl Deref for Name {
...
+        let mut world = World::new();
+        let e1 = world.spawn_empty().id();
+        let name = Name::new("MyName");
+        let e2 = world.spawn(name.clone()).id();
+        let mut query = world.query::<DebugName>();
+        let d1 = query.get(&world, e1).unwrap();
+        let d2 = query.get(&world, e2).unwrap();
+        // DebugName Display for entities without a Name should be {index}v{generation}
+        assert_eq!(d1.to_string(), "0v1");
+        // DebugName Display for entiites with a Name should be the Name
```
Comment (andresovela) [inline crates/bevy_core/src/name.rs:223]:
```suggestion
        // DebugName Display for entities with a Name should be the Name
```

File: crates/bevy_core/src/name.rs (hunk 3 above)
Comment (gagnus) [reply crates/bevy_core/src/name.rs]: Thanks for this, have updated the code if you could check its what you meant.

Comment (janhohenheim) [reply crates/bevy_core/src/name.rs]: Perfect, thanks :)

**Review by janhohenheim — APPROVED** [review]

**Comment by alice-i-cecile** [comment]
@eidloi, does this solve https://github.com/bevyengine/bevy/issues/13980 for you?

**Comment by eidloi** [comment]
> @eidloi, does this solve #13980 for you?

No, because it needs querying an additional component which may not be possible in some contexts. I mean it's not a bad thing to have but I find it somewhat redundant.

**Comment by gagnus** [comment]
Part of the original change was to make Display on an entity simpler (get rid of the redundant bit), but that change was I think more controversial than this. I’d support doing that, so this change is just a switch between two displays.

**Comment by gagnus** [comment]
@alice-i-cecile will this automatically go into 0.15 or can it go into an update to 0.14?

**Comment by alice-i-cecile** [comment]
Looks like this will be in 0.15 :)

**Pull request merged**